### ISSUE TITLE
[REVIEW]: 21cmFAST v3: A Python-integrated C code for generating 3D realizations of the cosmic 21cm signal.

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2582]
**Submitting author:** @steven-murray (<a href="http://orcid.org/0000-0003-3059-3823">Steven Murray</a>)
**Repository:** <a href="https://github.com/21cmFAST/21cmFAST" target ="_blank">https://github.com/21cmFAST/21cmFAST</a>
**Version:** v3.0.3
**Editor:** @dfm
**Reviewer:** @sambit-giri, @sultan-hassan
**Archive:** <a href="https://doi.org/10.5281/zenodo.4107189" target="_blank">10.5281/zenodo.4107189</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/56b5f5228d04377a9d612b30ce82b447/status.svg)](https://joss.theoj.org/papers/56b5f5228d04377a9d612b30ce82b447)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/56b5f5228d04377a9d612b30ce82b447"><img src="https://joss.theoj.org/papers/56b5f5228d04377a9d612b30ce82b447/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/56b5f5228d04377a9d612b30ce82b447/status.svg)](https://joss.theoj.org/papers/56b5f5228d04377a9d612b30ce82b447)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/21cmFAST/21cmFAST" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@sambit-giri & @sultanier, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @dfm know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/21cmFAST/21cmFAST">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@steven-murray) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/21cmFAST/21cmFAST">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@steven-murray) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @sambit-giri, @sultanier it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
PDF failed to compile for issue #2582 with the following error: 

 Can't find any papers to compile :-(

**Comment by dfm** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf)

**Comment by steven-murray** [comment]
Thanks again @sambit-giri and @sultanier for reviewing! And @dfm for editing.

**Comment by sambit-giri** [comment]
Hi @dfm 
I am unable to edit the reviewer's checklist. I don't remember if I had accepted the invitation. I apologise for it. When I try to do it now, I get the following message:
"Sorry, we couldn't find that repository invitation. It is possible that the invitation was revoked or that you are not logged into the invited account."

**Comment by danielskatz** [comment]
@whedon re-invite @sambit-giri as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@sambit-giri please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by sambit-giri** [comment]
Hi @steven-murray 
Here are a few comments on the software paper:

1. In paragraph 1, you say “Between these extremes lies an especially versatile middle-ground: fast semi-numerical models that approximate the full 3D evolution of the relevant fields: density, velocity, temperature, ionization, and radiation.” It is not clear which radiation you are talking about.
2. In paragraph 1, you say  “...so much so that they can be used to produce thousands of realizations on scales comparable to those observable by upcoming telescopes, in order....”. Can you mention the observables and few upcoming telescopes? Do you just mean the 21-cm signal and the upcoming radio telescopes? 21cmFast has also been used to put constraints based on other observations, such as quasar spectra and CMB observations. What about them?
3. The caption in Figure 1 does not describe the figure completely. The “component fields” are labelled with symbols in the figure. They have not been defined in the text and therefore are difficult to understand.
4. In paragraph 4, you say “This paper presents 21cmFAST v3+, which enables these essential new features. While keeping the same core functionality of previous versions of 21cmFAST, it…”. It is unclear which are the “new features” and which are the “same core functionality”.
5. There is an error in your citation in the fourth line of the list in paragraph 6.
6. In paragraph 7, I suggest you cite some example papers about milli-charged dark matter models and forward-modelled CMB auxiliary data.
7. Please cite the 21CMMC paper.
8. Please add a citation for “CLASS Boltzmann code”.
9. In Figure 3, please mention that ‘xH’ is the mass/volume average neutral fraction.

**Comment by dfm** [comment]
@sultanier: I just wanted to ping you to make sure that this is still on your radar. Let me know if you have any questions as you go through the checklist!

**Comment by sultanier** [comment]
thanks @dfm will do this asap.

**Comment by qyx268** [comment]
Hey @sambit-giri, thanks for reviewing this! I could answer the following points. @steven-murray Could you please add these in the paper?

> 3. The caption in Figure 1 does not describe the figure completely. The “component fields” are labelled with symbols in the figure. They have not been defined in the text and therefore are difficult to understand.

Lightcones for the reference model including, from left to right:
 			(i) overdensity ($\Delta_{\rm cell}$);
 			(ii) {\lya} ($J_{\alpha, {\rm eff}}{\times}\rm s^{-1}Hz^{-1}cm^{-2}sr^{-1}$);
 			(iii) LW ($J^{\rm 21}_{\rm LW,eff}{\times}{\rm 10^{{-}21}{\rm erg\ s^{-1}\ Hz^{-1}\ cm^{-2}\ sr^{-1}}}$);
 			(iv) X-ray heating ($\varepsilon_{X}$ in units of $k_\mathrm{B} {\rm K Gyr}^{-1}$);
 			(v) locally averaged UVB ($\bar{\Gamma}_{\rm ion}$ in units of $10^{-12}\rm{s}^{-1}$);
 			(vi) critical halo mass for star formation in ACGs ($M_{\rm crit}^{\rm atom}$/${\msol}$);
 			(vii) critical halo mass for star formation in MCGs ($M_{\rm crit}^{\rm mol}$/${\msol}$);
 			(viii) cumulative number of recombinations per baryon ($n_{\rm rec}$);
 			(ix) neutral hydrogen fraction ($x_{\hone}$); and
 			(x) the 21-cm brightness temperature ($\delta T_{\rm b}$ in units of ${\rm mK}$).\newline
 			{(A high-resolution version of this figure is available at \url{http://homepage.sns.it/mesinger/Media/light-cones_minihalo.png}.})


> 5. There is an error in your citation in the fourth line of the list in paragraph 6.

```
@ARTICLE{2020MNRAS.495..123Q,
       author = {{Qin}, Yuxiang and {Mesinger}, Andrei and {Park}, Jaehong and
         {Greig}, Bradley and {Mu{\~n}oz}, Julian B.},
        title = "{A tale of two sites - I. Inferring the properties of minihalo-hosted galaxies from current observations}",
      journal = {\mnras},
     keywords = {galaxies: high-redshift, intergalactic medium, dark ages, reionization, first stars, diffuse radiation, early Universe, cosmology: theory, Astrophysics - Cosmology and Nongalactic Astrophysics, Astrophysics - Astrophysics of Galaxies},
         year = 2020,
        month = apr,
       volume = {495},
       number = {1},
        pages = {123-140},
          doi = {10.1093/mnras/staa1131},
archivePrefix = {arXiv},
       eprint = {2003.04442},
 primaryClass = {astro-ph.CO},
       adsurl = {https://ui.adsabs.harvard.edu/abs/2020MNRAS.495..123Q},
      adsnote = {Provided by the SAO/NASA Astrophysics Data System}
}
```
> 8. Please add a citation for “CLASS Boltzmann code”.

```
@ARTICLE{Lesgourgues2011arXiv1104.2932L,
	author = {{Lesgourgues}, Julien},
	title = "{The Cosmic Linear Anisotropy Solving System (CLASS) I: Overview}",
	journal = {arXiv e-prints},
	keywords = {Astrophysics - Instrumentation and Methods for Astrophysics, Astrophysics - Cosmology and Extragalactic Astrophysics},
	year = 2011,
	month = apr,
	eid = {arXiv:1104.2932},
	pages = {arXiv:1104.2932},
	archivePrefix = {arXiv},
	eprint = {1104.2932},
	primaryClass = {astro-ph.IM},
	adsurl = {https://ui.adsabs.harvard.edu/abs/2011arXiv1104.2932L},
	adsnote = {Provided by the SAO/NASA Astrophysics Data System}
}
```

**Comment by steven-murray** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #2582 with the following error: 

 ORCID looks to be the wrong length
/app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-364ded062842/lib/whedon.rb:153:in `block in check_orcids': Problem with ORCID (https://orcid.org/0000-0003-1443-3483) for Catherine A. Watkinson (RuntimeError)
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-364ded062842/lib/whedon.rb:151:in `each'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-364ded062842/lib/whedon.rb:151:in `check_orcids'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-364ded062842/lib/whedon.rb:88:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-364ded062842/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-364ded062842/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-364ded062842/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-364ded062842/bin/whedon:119:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by steven-murray** [comment]
@sambit-giri: I've updated the paper trying to address your points (see https://github.com/21cmfast/21cmFAST/pull/105/commits/371e1d8d502cec94408e4c96a423f4e1930ecf40 for the diff).

In brief:


    

> 1. In paragraph 1, you say “Between these extremes lies an especially versatile middle-ground: fast semi-numerical models that approximate the full 3D evolution of the relevant fields: density, velocity, temperature, ionization, and radiation.” It is not clear which radiation you are talking about.

I've added that it is Lyman-alpha and 21-cm radiation.

> 2. In paragraph 1, you say “...so much so that they can be used to produce thousands of realizations on scales comparable to those observable by upcoming telescopes, in order....”. Can you mention the observables and few upcoming telescopes? Do you just mean the 21-cm signal and the upcoming radio telescopes? 21cmFast has also been used to put constraints based on other observations, such as quasar spectra and CMB observations. What about them?

I've restricted it to low-frequency radio telescopes. While other observables are also useful in constraining astrophysics using 21cmFAST, I wanted to focus on the primary motivation.

> 3. The caption in Figure 1 does not describe the figure completely. The “component fields” are labelled with symbols in the figure. They have not been defined in the text and therefore are difficult to understand.

Thanks for picking this up, I've added captions to all the figures, and this one describes the components.

> 4. In paragraph 4, you say “This paper presents 21cmFAST v3+, which enables these essential new features. While keeping the same core functionality of previous versions of 21cmFAST, it…”. It is unclear which are the “new features” and which are the “same core functionality”.

I've changed the wording here to be the "essential guiding principals", and it refers back to the end of the previous paragraph. It then goes on to describe the new features.

> 5. There is an error in your citation in the fourth line of the list in paragraph 6.

Fixed, thanks!

> 6. In paragraph 7, I suggest you cite some example papers about milli-charged dark matter models and forward-modelled CMB auxiliary data.

Added, thanks!

> 7. Please cite the 21CMMC paper.
> 8. Please add a citation for “CLASS Boltzmann code”.
> 9. In Figure 3, please mention that ‘xH’ is the mass/volume average neutral fraction.

All done.

**Comment by steven-murray** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: :point_left:

**Comment by steven-murray** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: :point_left:

**Comment by sultanier** [comment]
@whedon  I can't mark the checklist, could you re-invite me again? thanks.

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by sultanier** [comment]
@whedon re-invite @sultanier as reviewer

**Comment by whedon** [comment]
I'm sorry @sultanier, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by dfm** [comment]
@whedon re-invite @sultanier as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@sultanier please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by steven-murray** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: :point_left:

**Comment by sambit-giri** [comment]
Hi @steven-murray 

Can you make the GitHub repository self-sustained? You can add a short summary of what 21cmFast can do. It will be helpful for new users. You should also add some of the easiest installation method (for more details you provide a link to your readthedocs installation page), instructions about automated tests post-installation and community guidelines (contribution, raising issue etc).

I installed the package in MacOSX. The instructions in the readthedocs page was not enough. I had to refer to the GitHub issue you mention. Is it possible to improve the instructions by summarising a few of the important information discussed in the issue. Also, "conda install -c conda-forge 21cmFAST" worked better for me. Please mention this method in the readthedocs page and GitHub readme.

**Comment by sambit-giri** [comment]
@steven-murray 

My automated tests failed. I am attaching a file containing the output I received while running `python -m pytest`.

[pytest_output.txt](https://github.com/openjournals/joss-reviews/files/5286637/pytest_output.txt)

**Comment by sambit-giri** [comment]
@steven-murray 
Thanks for updating the software paper.

Just a final comment on that. The link given for Figure 1 is wrong. There is a hyphen in `light-cones` in the link given, but I see the high-resolution image at http://homepage.sns.it/mesinger/Media/lightcones_minihalo.png

**Comment by steven-murray** [comment]
Hi @sambit-giri, thanks for the very useful suggestions!


> Can you make the GitHub repository self-sustained? You can add a short summary of what 21cmFast can do. It will be helpful for new users. 

I've made these changes in https://github.com/21cmfast/21cmFAST/pull/172 (hopefully merged soon). The description is short but hopefully helpful.

> You should also add some of the easiest installation method (for more details you provide a link to your readthedocs installation page), 

Done in the above PR as well.

> instructions about automated tests post-installation and community guidelines (contribution, raising issue etc).

Running automated tests post-installation is only possible if one clones the repo itself and installs that way. We expect that to happen only for those who want to be developers, in which case the developer documentation outlines how to run the tests.
Also, we already have community guidelines in the organization's `.github` repo, which should show up when a new user tries to create an issue or PR. Agreed that it is less easy to find those instructions manually this way, but it means we have the same instructions for 21cmMC without redundant files. 

> I installed the package in MacOSX. The instructions in the readthedocs page was not enough. I had to refer to the GitHub issue you mention. Is it possible to improve the instructions by summarising a few of the important information discussed in the issue. Also, conda install -c conda-forge 21cmFAST worked better for me. Please mention this method in the readthedocs page and GitHub readme.

The installation page on RTD has been updated to include the simpler conda install (however, I've just noticed that the RTD build is failing, hence you didn't see it -- will fix that ASAP). Installing on MacOS without `conda` is just hard, and it would be difficult to summarise the different cases succinctly. I feel like referencing the issue is the best we can do for now. 

Concerning your failed tests -- can you inform me of which version you're testing (master branch or otherwise?). CI is passing all tests on master currently (on MacOS as well), so it is strange that they fail for you.

I've also updated the link in the paper, thanks!

**Comment by sultanier** [comment]
Hi @steven-murray, seems like a very nice package, congrats! 

 - The installation went great as described, however just small suggestion to add might help new users 
to computing clusters that these dependencies like gsl, fftw..etc might already be installed so you might 
add: use e.g. module load gsl ....

- I like the CLI interactive mode with the 21cmfast command but I was unable to run it.
Could you give a complete example of how to use it? Also would it be possible to add the whole parameter list here and their meanings? because the config file isn't complete ~/.21cmfast/runconfig_example.yml, and should one copies the config file to another location  or has to be there ..etc? Would be nice to include an example of a complete config file with all parameters and their definitions so a new user would know how to specify them, and where to store the generated simulation output.  I see in the following page you have tutorial of how to use the package within python, but some users, like me, might prefer the interactive mode with the 21cmfast command.

**Comment by sultanier** [comment]
@steven-murray, how do we see all options for the 'kind' argument in the coeval_sliceplot, I couldn't find this in the API. the actual spelling might be different, also they aren't in the flag_options: coeval8.flag_options:
FlagOptions(INHOMO_RECO:False, M_MIN_in_Mass:False, PHOTON_CONS:False, SUBCELL_RSD:False, USE_HALO_FIELD:False, USE_MASS_DEPENDENT_ZETA:False, USE_MINI_HALOS:False, USE_TS_FLUCT:False)

**Comment by sultanier** [comment]
@steven-murray the paper reads very nice, I think it misses how fast is this package compared to just running 21cmFAST directly? how the linking between python and C is achieved (OS module, Ctypes, Cython...etc)? The community would be interested to know the time spent for few examples: e.g. to to run a lightcone from z=20 to z=6 for box size 100 Mpc the code spends ...etc this is also relevant for users to have a rough estimate if they want to run MCMC with that. Otherwise looking good congrats again.

**Comment by steven-murray** [comment]
Thanks @sultanier for the very useful comments on the code! I've tried to address most of your comments in this PR: https://github.com/21cmfast/21cmFAST/pull/174



> The installation went great as described, however just small suggestion to add might help new users
   to computing clusters that these dependencies like gsl, fftw..etc might already be installed so you might
   add: use e.g. module load gsl ....

Thanks, I added a `HPC` section to the installation docs.

> I like the CLI interactive mode with the 21cmfast command but I was unable to run it.
> Could you give a complete example of how to use it? Also would it be possible to add the whole parameter list here and their        meanings? because the config file isn't complete ~/.21cmfast/runconfig_example.yml, and should one copies the config file to another location or has to be there ..etc? Would be nice to include an example of a complete config file with all parameters and their definitions so a new user would know how to specify them, and where to store the generated simulation output. I see in the following page you have tutorial of how to use the package within python, but some users, like me, might prefer the interactive mode with the 21cmfast command.

Firstly, could you let me know what went wrong for you with the CLI command? 

I've updated the documentation in the README to have a quick guide on using the CLI. It references the user to https://21cmfast.readthedocs.io/en/latest/reference/_autosummary/py21cmfast.inputs.html to get a list of all the possible inputs and their documentation. I've also put this URL into the `runconfig_example.yml`. The config file location is specified with `--config=FILE` on the CLI, which is also now documented in the README. I would argue against including a COMPLETE config, as it would be doubling-up the documentation of the code itself (and may get out of date). As for storing the generated output, the README now has some description of this, and you are able to save the (coeval or lightcone) file explicitly. I hope these things make it easier to use!



> could you add in the beginning of section: Running and Plotting Coeval Cubes after "Clear the cache so that we get the same results for the notebook every time (don’t worry about this for now Also, set the default output directory to _cache/:" create dir, mkdir '_cache' where you want to run this jupyternotebook, otherwise it says _cache not found.

Done, thanks!

 > when you save files, is there anyway to make the saved files names more meaningful, for instance in this saved box name "BrightnessTemp_8da85e6e1d0e198f6a7f0f6b3af0a9a2_r12345.h5", I recognize that the last part in the files names is the seed number, is there any way for the user to specify the name? or the name can be changed to reflect other cosmo params?

The short answer is no, the automatically-saved cache files have a particular naming structure that is based on a hash of all the parameters that went into the simulation. We have tried to keep the most useful information in the filename (i.e. the kind of box it is, and the seed). Naming the file based on some subset of its parameters is not scalable, and would not do well as a cache. Of course, you can always copy whatever file you want to somewhere else on your filesystem and call it whatever you want! Also, we have made the `21cmfast query` tool, which lists each box and all the parameters that went into it, to make it a bit easier to know what each simulation really is. 

Finally, the higher level outputs (Coeval and Lightcone) *do* have a `.save()` method which allows you to save the box with any filename you want. They are not part of the caching system, so this is OK. 

> @steven-murray, how do we see all options for the 'kind' argument in the coeval_sliceplot, I couldn't find this in the API. the actual spelling might be different, also they aren't in the flag_options: coeval8.flag_options:
FlagOptions(INHOMO_RECO:False, M_MIN_in_Mass:False, PHOTON_CONS:False, SUBCELL_RSD:False, USE_HALO_FIELD:False, USE_MASS_DEPENDENT_ZETA:False, USE_MINI_HALOS:False, USE_TS_FLUCT:False)

This is actually an open issue (https://github.com/21cmfast/21cmFAST/issues/81). I've tried to partially address it in the PR I linked above. There is now a method that can be called on the `Coeval` class (you can do it directly, before having run a simulation): `Coeval.get_fields()`, that will list these kinds for you. I've also added this information to the `coeval_sliceplot` function docstring. It's not a full solution, as the output fields are not actually documented online, but I will leave that issue open to remind us to do this in the future.

> @steven-murray the paper reads very nice, I think it misses how fast is this package compared to just running 21cmFAST directly? how the linking between python and C is achieved (OS module, Ctypes, Cython...etc)? The community would be interested to know the time spent for few examples: e.g. to to run a lightcone from z=20 to z=6 for box size 100 Mpc the code spends ...etc this is also relevant for users to have a rough estimate if they want to run MCMC with that. Otherwise looking good congrats again.

This is a great suggestion. I will add another small section into the paper to describe this.

**Comment by sultanier** [comment]
thanks @steven-murray  

for the CLI command, I was't sure how to run it because there was no detailed config file and the required parameter list was not mentioned. I will try again once you add a detailed config file. Let me know once you finish adding/implementing these suggestions/comments, so I can double check them for the last time before finishing up my checklist. thanks.

**Comment by steven-murray** [comment]
@sultanier all the comments have been address in the current master (there are some formatting issues with the hyperlinks in the readme which are being addressed in a PR, but these shouldn't hamper your checks). The performance information to go into the paper itself is still a work in progress.

**Comment by sultanier** [comment]
thanks @steven-murray, 

I have finished checking all comments and everything looking good and working nicely! congrats!

One small fix is that these hyperlinks below for the config_example and inputs don't work (clicking on them doesn't directly take you to the files), but one can go to the repo and see these files. Just make sure these hyperlinks do actually work (if meant). I have completed all of my checklist but I am very curious to know the performance and speed for different examples, once you finish please generate the new pdf and let me know. Package is very beautifully made, congrats again!.

# hyperlinks.....
There is an example configuration file `here<user_data/runconfig_example.yml>`_ that you can build from. All input parameters are `documented here<https://21cmfast.readthedocs.io/en/latest/reference/_autosummary/py21cmfast.inputs.html>`_.

**Comment by steven-murray** [comment]
Thanks @sultanier! There is a PR currently to fix those hyperlinks (should be merged in the next couple of days). We'll have those performance metrics into the paper very soon as well.

**Comment by steven-murray** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #2582 with the following error: 

 Can't find any papers to compile :-(

**Comment by steven-murray** [comment]
@sultanier the updates to the hyperlinks have been merged. The benchmarks have also been added to the paper, but I can't seem to get it to compile. @dfm do you have any idea what could be going wrong there? The paper is definitely still there in the `joss-paper` branch...

@sambit-giri I'm not sure if all your questions/comments were addressed adequately?

**Comment by sambit-giri** [comment]
@steven-murray My questions/comments were addressed. I thought that I had checked all the points, which I noticed now that I hadn't. Sorry for the delay.

**Comment by sambit-giri** [comment]
@dfm I am done with the reviewing process.
Is there any other unfinished formality?

**Comment by dfm** [comment]
@sambit-giri and @sultanier: You're both all set! Thank you for your time and for your constructive reviews!!

@steven-murray: I'll see if I can figure out the whedon issues and get back to you with next steps.

**Comment by dfm** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #2582 with the following error: 

 Can't find any papers to compile :-(

**Comment by dfm** [comment]
Pinging @openjournals/dev to see if there are any thoughts about why @whedon can't find the paper. I can confirm that [whedon.theoj.org](https://whedon.theoj.org/) is able to find it. Thanks!

**Comment by arfon** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #2582 with the following error: 

 Can't find any papers to compile :-(

**Comment by arfon** [comment]
Weird, I need to dig into this more. Pls hold.

**Comment by dfm** [comment]
Thanks @arfon!

**Comment by sultan-hassan** [comment]
Great, looking forward to see the last version of the paper. congrats!

Sultan

p.s. just a note for your reference, I have changed my username to @sultan-hassan

**Comment by arfon** [comment]
OK, I've figured out what's going on. I've not quite figured out the 'proper' fix here but I have a short-term solution. **tl;dr** – delete (or rename) the `joss-paper` folder in your default (`master`) branch and everything should be fine. 

If you're interested, here's what's going on:

The issue here is that we're asking Whedon to act on the `joss-paper` branch but there's also a folder named `joss-paper` in the default (`master`) branch.

This confuses Git<sup>1</sup>, and I thought that [this commit](https://github.com/openjournals/whedon-api/commit/6c8d7d9127c6483e7c48c1097ff17d1876e77711) would address the issue (it does locally with my local Git version) but Whedon's Git is older and this flag doesn't seem to work there.

Ultimately I need to spend more time debugging this but I'm out of time this evening and don't want to block this further.

<sup>1</sup> https://stackoverflow.com/questions/25322335/git-change-branch-when-file-of-same-name-is-present

**Comment by steven-murray** [comment]
I'll just merge the PR and we can build off master :-)

**Comment by steven-murray** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: :point_left:

**Comment by sultan-hassan** [comment]
Hi @dfm, I have chnaged my github username, is it possible to update that in the paper too? thanks.

**Comment by dfm** [comment]
@sultan-hassan: yes thanks for the heads up! I'll fix it in the process of the final edit.

**Comment by steven-murray** [comment]
@whedon generate pdf from branch joss-paper-fixes

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper-fixes. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: :point_left:

**Comment by sambit-giri** [comment]
Congrats to all the authors!

**Comment by sambit-giri** [comment]
@dfm I noticed that the author's list of the paper has been modified. I have worked with Catherine Watkinson in a project where both of us were primary authors (https://ui.adsabs.harvard.edu/abs/2019MNRAS.482.2653W/abstract). Does this violate the JOSS conflict of interest policy?

**Comment by dfm** [comment]
@sambit-giri: Thanks for the heads-up! I'll check in with the editorial board and get back to you. Thanks!

**Comment by dfm** [comment]
@steven-murray: I've opened a PR to the 21cmfast repo with a few minor formatting edits to the paper. Please take a look (and preferably accept) those changes and then go through the following steps:

1. I see now that there is also a `joss-paper-fixes` branch that will have one conflict with my PR. Please resolve those issues and confirm which branch the final submission will live on.
2. Get whedon to compile the paper and then read through the proofs carefully (especially checking things like author names and affiliations, but also general typos that we might have all missed) to make sure that you're happy. It's much harder to change things later! I believe that @sultan-hassan's username should be updated in the new version of the paper, but we should also check that.
3. Bump the version number and report the version number back here.
4. Create a tagged release and archive this release somewhere like Zenodo. The metadata of that archive (title and author list) **must exactly match the JOSS paper**. You can edit the metadata after creating the archive on most services. Report the DOI for the archive in this thread.

Let me know if you have any questions!

**Comment by dfm** [comment]
@sambit-giri: I've checked with the editorial team and we've agreed that this does not constitute a conflict-of-interest in this case since you had already done most of your review and since it has been noted here. If @steven-murray or any of the co-authors have any concerns, please let me know here or offline, but otherwise we'll proceed. Thanks!

**Comment by steven-murray** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: :point_left:

**Comment by steven-murray** [comment]
OK @dfm, I've merged the paper fixes, and the proof looks good.

As for the version update -- does that need to be a new version? Can we point to the last published version (3.0.2)? If not that's fine, but may be a day or so before we can release a new version.

**Comment by dfm** [comment]
The version for the paper should represent the current state of the code base and docs. It looks to me like there have been a fair number of updates since 3.0.2 (July 2020) so I think you should make a new one. If you think otherwise, please let me know!

**Comment by steven-murray** [comment]
You're right, @dfm. Made a new version.

New version is v3.0.3 and DOI is 10.5281/zenodo.4107189.

**Comment by dfm** [comment]
Thanks! Can you edit the Zenodo meta data so that the title is "21cmFAST v3: A Python-integrated C code for generating 3D realizations of the cosmic 21cm signal"?

**Comment by steven-murray** [comment]
:facepalm: done.

**Comment by dfm** [comment]
@whedon set 10.5281/zenodo.4107189 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4107189" target="_blank">10.5281/zenodo.4107189</a> is the archive.

**Comment by dfm** [comment]
@whedon set v3.0.3 as version

**Comment by whedon** [comment]
OK. v3.0.3 is the version.

**Comment by dfm** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1086/521806 is OK
- 10.1111/j.1365-2966.2010.17731.x is OK
- 10.1086/423025 is OK
- 10.1046/j.1365-8711.2002.04999.x is OK
- 10.1103/PhysRevLett.123.131301 is OK
- 10.1103/PhysRevD.100.063538 is OK
- 10.1093/mnras/staa1131 is OK
- 10.1103/PhysRevLett.121.121301 is OK
- 10.1093/mnras/stv571 is OK

MISSING DOIs

- 10.1093/mnras/staa2797 may be a valid DOI for title: Reionization inference from the CMB optical depth and E-mode polarization power spectra

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1831

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1831, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by dfm** [comment]
@steven-murray: I see now that I missed that updated reference. It looks like https://ui.adsabs.harvard.edu/abs/2020MNRAS.499..550Q/abstract has now been published so can you update the bibtex entry to the published version?

EiC: I'll remove the `recommend-accept` label until @steven-murray fixes that. Sorry all about the false alarm!

**Comment by steven-murray** [comment]
Thanks for spotting that @dfm. It's updated in master now.

**Comment by dfm** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1086/521806 is OK
- 10.1111/j.1365-2966.2010.17731.x is OK
- 10.1086/423025 is OK
- 10.1046/j.1365-8711.2002.04999.x is OK
- 10.1103/PhysRevLett.123.131301 is OK
- 10.1103/PhysRevD.100.063538 is OK
- 10.1093/mnras/staa1131 is OK
- 10.1103/PhysRevLett.121.121301 is OK
- 10.1093/mnras/staa2797 is OK
- 10.1093/mnras/stv571 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by dfm** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: :point_left:

**Comment by dfm** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1086/521806 is OK
- 10.1111/j.1365-2966.2010.17731.x is OK
- 10.1086/423025 is OK
- 10.1046/j.1365-8711.2002.04999.x is OK
- 10.1103/PhysRevLett.123.131301 is OK
- 10.1103/PhysRevD.100.063538 is OK
- 10.1093/mnras/staa1131 is OK
- 10.1103/PhysRevLett.121.121301 is OK
- 10.1093/mnras/staa2797 is OK
- 10.1093/mnras/stv571 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1834

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1834, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kyleniemeyer** [comment]
Hi @steven-murray, I am the EIC on duty this week, doing some final checks before publishing.

I noticed that in the paper, journal names are not appearing for your references. Can you replace all the macros in your .bib file (e.g., `\apj`) with the spelled-out journal names?

**Comment by steven-murray** [comment]
Thanks @kyleniemeyer . This is done now and in master.

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02582/joss.02582/10.21105.joss.02582.pdf) :page_facing_up: :point_left:

**Comment by kyleniemeyer** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1086/521806 is OK
- 10.1111/j.1365-2966.2010.17731.x is OK
- 10.1086/423025 is OK
- 10.1046/j.1365-8711.2002.04999.x is OK
- 10.1103/PhysRevLett.123.131301 is OK
- 10.1103/PhysRevD.100.063538 is OK
- 10.1093/mnras/staa1131 is OK
- 10.1103/PhysRevLett.121.121301 is OK
- 10.1093/mnras/staa2797 is OK
- 10.1093/mnras/stv571 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1840

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1840, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1319300599763095552) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1841
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02582](https://doi.org/10.21105/joss.02582)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congrats @steven-murray on your article's publication in JOSS!

Many thanks to @sambit-giri and @sultan-hassan for reviewing this, and @dfm for editing.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02582/status.svg)](https://doi.org/10.21105/joss.02582)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02582">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02582/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02582/status.svg
   :target: https://doi.org/10.21105/joss.02582
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02582/status.svg)](https://doi.org/10.21105/joss.02582)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by steven-murray** [comment]
Thanks @dfm , @sultan-hassan, @sambit-giri and @kyleniemeyer!
